packages/@aws-containers/aws-ecs-builder/lib/extensions/extension-interfaces.ts

```diff
@@ -0,0 +1,218 @@
+   */
+  public addHooks() { } // tslint:disable-line
+
+  /**
+   * This hook allows another service extension to register a mutating hook for
+   * changing the primary container of this extension. This is primarily used
+   * for the application extension. For example the Firelens extension wants to
+   * be able to modify the settings of the application container to
+   * route logs through Firelens.
+   * @param hook
+   */
+  public addContainerMutatingHook(hook: ContainerMutatingHook) {
+    this.containerMutatingHooks.push(hook);
+  }
+
+  /**
+   * This is a hook which allows extensions to modify the settings of the
+   * task definition prior to it being created. For example App Mesh
+   * extension needs to configure an Envoy proxy in the task definition,
+   * or the application extension wants to set the overall resource for
+   * the task.
+   * @param props - Properties of the task definition to be created
+   */
+  public mutateTaskDefinitionProps(props: ecs.TaskDefinitionProps) {
+    return {
+      ...props,
+    } as ecs.TaskDefinitionProps;
+  }
+
+  /**
+   * A hook that is called for each extension adhead of time to
+   * let it do any initial setup, such as creating resources in
+   * advance.
+   * @param parent - The parent service which this extension has been added to
+   * @param scope - The scope that this extension should create resources in
+   */
+  public prehook(parent: Service, scope: cdk.Construct) {
+    this.parentService = parent;
+    this.scope = scope;
+  }
+
+  /**
+   * Once the task definition is created, this hook is called for each
+   * extension to give it a chance to add containers to the task definition,
+   * change the task definition's role to add permissions, etc
+   * @param taskDefinition - The created task definition to add containers to
+   */
+  public useTaskDefinition(taskDefinition: ecs.TaskDefinition) {
+    taskDefinition = taskDefinition;
+  }
+
+  /**
+   * Once all containers are added to the task definition this hook is
+   * called for each extension to give it a chance to bake its dependency
+   * graph so that its container starts in the right order based on the
+   * other extensions that were enabled
+   */
+  public bakeContainerDependencies() {
+    return;
+  }
+
+  /**
+   * Prior to launching the task definition as a service this hook
+   * is called on each extension to give it a chance to mutate the properties
+   * of the service to be created
+   * @param props - The service properties to mutate
+   */
+  public mutateServiceProps(props: ServiceBuild) {
+    return {
+      ...props,
+    } as ServiceBuild;
+  }
+
+  /**
+   * When this hook is implemented by extension it allows the extension
+   * to use the service which has been created. It is generally used to
+   * create any final resources which might depend on the service itself
+   * @param service - The generated service
+   */
+  public useService(service: ecs.Ec2Service | ecs.FargateService) {
```

Comment:
I also don't love the `use` verb here, but not sure if I have a better suggestion.